File: src/swift/4_BehavioralPatterns/9_State/Exercise.swift
import Foundation
import XCTest

class CombinationLock
{
  private let combination: [Int]
  var status = ""
  private var digitsEntered = 0
  private var failed = false

  init(_ combination: [Int])
  {
    self.combination = combination
    reset()
  }

  private func reset()
  {
    status = "LOCKED"
    digitsEntered = 0
    failed = false
  }

  func enterDigit(_ digit: Int)
  {
    if (status == "LOCKED") { status = "" }
    status += String(digit)
    if combination[digitsEntered] != digit
    {
      failed = true
    }
    digitsEntered += 1

    if digitsEntered == combination.count
    {
      status = (failed ? "ERROR" : "OPEN")
    }
  }
}

// ===

class UMBaseTestCase : XCTestCase {}

class Evaluate: UMBaseTestCase
{
  func testSuccess()
  {
    let cl = CombinationLock([1,2,3,4,5])
    XCTAssertEqual("LOCKED", cl.status)
    cl.enterDigit(1)
    XCTAssertEqual("1", cl.status)
    cl.enterDigit(2)
    XCTAssertEqual("12", cl.status)
    cl.enterDigit(3)
    XCTAssertEqual("123", cl.status)
    cl.enterDigit(4)
    XCTAssertEqual("1234", cl.status)
    cl.enterDigit(5)
    XCTAssertEqual("OPEN", cl.status)
  }

  func testFailure()
  {
    let cl = CombinationLock([1,2,3])
    XCTAssertEqual("LOCKED", cl.status)
    cl.enterDigit(1)
    XCTAssertEqual("1", cl.status)
    cl.enterDigit(2)
    XCTAssertEqual("12", cl.status)
    cl.enterDigit(5)
    XCTAssertEqual("ERROR", cl.status)
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("testSuccess", testSuccess),
      ("testFailure", testFailure),
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/4_BehavioralPatterns/8_State/exercise.py
import unittest


class CombinationLock:
    def __init__(self, combination):
        self.combination = combination
        self.reset()

    def reset(self):
        self.status = 'LOCKED'
        self.digits_entered = 0
        self.failed = False

    def enter_digit(self, digit):
        if self.status == 'LOCKED':
            self.status = ''
        self.status += str(digit)
        if self.combination[self.digits_entered] != digit:
            self.failed = True
        self.digits_entered += 1

        if self.digits_entered == len(self.combination):
            self.status = 'ERROR' if self.failed else 'OPEN'


class FirstTestSuite(unittest.TestCase):
    def test_success(self):
        cl = CombinationLock([1, 2, 3, 4, 5])
        self.assertEqual('LOCKED', cl.status)
        cl.enter_digit(1)
        self.assertEqual('1', cl.status)
        cl.enter_digit(2)
        self.assertEqual('12', cl.status)
        cl.enter_digit(3)
        self.assertEqual('123', cl.status)
        cl.enter_digit(4)
        self.assertEqual('1234', cl.status)
        cl.enter_digit(5)
        self.assertEqual('OPEN', cl.status)

    def test_failure(self):
        cl = CombinationLock([1, 2, 3])
        self.assertEqual('LOCKED', cl.status)
        cl.enter_digit(1)
        self.assertEqual('1', cl.status)
        cl.enter_digit(2)
        self.assertEqual('12', cl.status)
        cl.enter_digit(5)
        self.assertEqual('ERROR', cl.status)
